Bugfixes and improvements on sparse

ezyang opened the pull request:
I extracted out the commits that had nothing to do with sparse_mask/sparse_copy from #1471 and I am resubmitting them here. These patches have already been reviewed by @apaszke.

ezyang committed:
Sparse tensor clone() preserves coalescedness.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add some docs about storage->Size.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Test clone preserves uncoalescedness if it wasn't coalesced.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
More efficient nnz test in spcadd.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Abort on known bug (#1521) for spcadd on non-coalesced.

It's better to error than to silently give wrong results.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on torch/lib/THCS/generic/THCSTensorMath.cu:
```diff
@@ -217,21 +217,25 @@ void THCSTensor_(hspmm)(THCState *state, THCSTensor *r_, real alpha, THCSTensor
+  const ptrdiff_t nnz = THCSTensor_(nnz)(state, sparse);
+  if (nnz == 0) {
+    THCTensor_(resizeAs)(state, r_, dense);
+    THCTensor_(copy)(state, r_, dense);
```
this will perform an unnecessary copy if `r_ == dense`

ezyang replied:
It looks like I got tricked, because `THCSTensor_(copy)` will bail out if `r_ == dense`, but it doesn't look like `THCTensor_(copy)` will. Should we also make `THCTensor_(copy)` do this test?

ezyang replied:
I suggest:

{{{
diff --git a/torch/lib/TH/generic/THTensorCopy.c b/torch/lib/TH/generic/THTensorCopy.c
index 71ccfdd..6e08ba6 100644
--- a/torch/lib/TH/generic/THTensorCopy.c
+++ b/torch/lib/TH/generic/THTensorCopy.c
@@ -68,7 +68,9 @@ void THTensor_(copyTranspose)(THTensor *tensor, THTensor *src) {
 
 void THTensor_(copy)(THTensor *tensor, THTensor *src)
 {
-  if (THTensor_(isContiguous)(tensor) && THTensor_(isContiguous)(src) && THTensor_(nElement)(tensor) == THTensor_(nElement)(src)) {
+  if (tensor == src) {
+    // Nothing to do
+  } else if (THTensor_(isContiguous)(tensor) && THTensor_(isContiguous)(src) && THTensor_(nElement)(tensor) == THTensor_(nElement)(src)) {
     real *sp = THTensor_(data)(src);
     real *rp = THTensor_(data)(tensor);
     ptrdiff_t sz = THTensor_(nElement)(tensor);
diff --git a/torch/lib/THC/generic/THCTensorCopy.cu b/torch/lib/THC/generic/THCTensorCopy.cu
index 4198025..f99ea44 100644
--- a/torch/lib/THC/generic/THCTensorCopy.cu
+++ b/torch/lib/THC/generic/THCTensorCopy.cu
@@ -4,7 +4,9 @@
 
 THC_API void
 THCTensor_(copy)(THCState* state, THCTensor* dst, THCTensor* src) {
-  THC_copyTensor<THCTensor, THCTensor>(state, dst, src);
+  if (dst != src) {
+    THC_copyTensor<THCTensor, THCTensor>(state, dst, src);
+  }
 }
 
 THC_API void
}}}

apaszke replied:
Looks good. My only nit is that `if (tensor == src) return;` is enough to say "nothing to do". Should be good to go once you add these changes

ezyang committed:
Short-circuit copy if src and dest are equal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Done!

ezyang committed:
Add comment about new implementation being CPU-only.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Do NOT clone indices/values passed to sparse tensor by default.

Fixes #1782.

The default operation should be cheap: user can always choose to
explicitly make a copy on the way in.  Note that this is a
BACKWARDS COMPATIBILITY BREAKING change.  However, we DO create
a new tensor wrapper (so we are not affected by subsequent
size changes, etc.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add comment stating nDenseTensors != nTensors in checkGPU.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Enforce storage is not NULL invariant for sparse tensors.

Fixes #1783.

There is an undocumented invariant in PyTorch that we should
try to avoid having storage == NULL as much as possible (even
though Torch supports it.)  This commit properly documents the
invariant, and fixes a bug in sparse where the invariant was
not respected.  This now means that sparse tensors now correctly
remember what GPU they are associated with.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Support device on sparse tensor constructor, assert values/indices on same device.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
I pushed some more commits improving sparse's behavior in a multi-GPU setting. There is also a big change: we no longer clone by default when constructing a sparse from an index and value tensor. CC @apaszke, @martinraison, @adamlerer.

apaszke reviewed (commented):
This looks good, but I think we should do a shallow clone (new tensor, same storage) instead of just taking `indices` and `values`.

ezyang commented:
Shallow clone is now implemented.

soumith commented:
this is now merged into master!
